module/x/gravity/abci.go

```diff
@@ -275,135 +317,67 @@ func cleanupTimedOutContractCallTxs(ctx sdk.Context, k keeper.Keeper) {
-								sdk.NewAttribute(slashingtypes.AttributeKeyPower, fmt.Sprintf("%d", power)),
-							),
-						)
-					}
-				}
+
+		// Slash bonded validators who didn't sign outgoing txs
+		for _, validator := range bondedValidators {
+			signingStartHeight, signingInfoExists := k.GetValidatorSlashingCriteria(ctx, validator)
+			eligibleSigner := signingInfoExists && (signingStartHeight < int64(otx.GetCosmosHeight()))
+			_, signedTx := signatures[validator.GetOperator().String()]
+
+			if eligibleSigner && !signedTx {
+				k.SlashAndJail(ctx, validator, types.AttributeMissingSignature)
 			}
 		}
 
+		// Slash unbonding validators who didn't sign new signer set txs only
 		if sstx, ok := otx.(*types.SignerSetTx); ok {
-			for _, valInfo := range unbondingValInfos {
-				// Only slash validators who joined after valset is created and they are
-				// unbonding and UNBOND_SLASHING_WINDOW didn't pass.
-				if valInfo.exist && valInfo.sigs.StartHeight < int64(sstx.Height) &&
-					valInfo.val.IsUnbonding() &&
-					sstx.Height < uint64(valInfo.val.UnbondingHeight)+params.UnbondSlashingSignerSetTxsWindow {
-					// check if validator has confirmed valset or not
-					if _, found := signatures[valInfo.val.GetOperator().String()]; !found {
-						if !valInfo.val.IsJailed() {
-							// TODO: Do we want to slash jailed validators?
-							power := valInfo.val.ConsensusPower(k.PowerReduction)
-							k.StakingKeeper.Slash(
-								ctx,
-								valInfo.cons,
-								ctx.BlockHeight(),
-								power,
-								params.SlashFractionSignerSetTx,
-							)
-							k.StakingKeeper.Jail(ctx, valInfo.cons)
-
-							ctx.EventManager().EmitEvent(
-								sdk.NewEvent(
-									slashingtypes.EventTypeSlash,
-									sdk.NewAttribute(slashingtypes.AttributeKeyAddress, valInfo.cons.String()),
-									sdk.NewAttribute(slashingtypes.AttributeKeyJailed, valInfo.cons.String()),
-									sdk.NewAttribute(slashingtypes.AttributeKeyReason, types.AttributeMissingBridgeBatchSig),
-									sdk.NewAttribute(slashingtypes.AttributeKeyPower, fmt.Sprintf("%d", power)),
-								),
-							)
-						}
-					}
+			for _, validator := range unbondingValidators {
+				signingStartHeight, signingInfoExists := k.GetValidatorSlashingCriteria(ctx, validator)
+				eligibleSigner := signingInfoExists && (signingStartHeight < int64(sstx.Height))
+				deadlinePassed := sstx.Height < uint64(validator.UnbondingHeight)+params.UnbondSlashingSignerSetTxsWindow
+				_, signedTx := signatures[validator.GetOperator().String()]
+
+				if eligibleSigner && validator.IsUnbonding() && deadlinePassed && !signedTx {
+					k.SlashAndJail(ctx, validator, types.AttributeMissingSignerSetSignature)
 				}
 			}
 		}
 
-		// then we set the latest slashed outgoing tx block
-		k.SetLastSlashedOutgoingTxBlockHeight(ctx, otx.GetCosmosHeight())
+		if otx.GetCosmosHeight() > k.GetLastSlashedOutgoingTxBlockHeight(ctx) {
```

Comment:
Out of curiosity, how could this possibly not be true, given that this is called once per block in `EndBlocker`? Are we expecting to see OTXs out of order?

Reply:
This block is at the end of a loop; the check ensures only setting the `LastSlashedOutgoingTxBlockHeight` once in case there are multiple slashings in the iteration. It's just to avoid needlessly setting the value multiple times to the same thing.